doc/thingpedia-nl-support.md

```diff
@@ -67,10 +154,80 @@ action share(in req status: String #_[prompt="What do you want to post?"])
+It has only one property: `text`. 
+E.g., `{ type="text",text="Works in ${industry}" }` (in [Tutorial 3](/doc/thingpedia-tutorial-linkedin.md)).
+
+####  Picture response: 
+A picture response shows a picture to the user. 
+It has only one property: `url`.
+E.g., `{ type: "picture", url: "${picture_url}" }` (in [Tutorial 3](/doc/thingpedia-tutorial-linkedin.md)).
+
+#### RDL response: 
+An RDL response returns a clickable with a title and a description, suitable for website 
+links and news articles.
+It has three properties: `webCallback` (the link), `displayTitle` (the title), 
+and `displayText` (the description, optional). 
+See [Tutorial 1](/doc/thingpedia-tutorial-nyt.md) for an example of this format type.
+
+#### Customized response
+If you need more control over the output, such as different output based 
+on results, you can choose `code` type and write Javascript code in the `code` property. 
+The code contains a [self-executing anonymous function](https://developer.mozilla.org/en-US/docs/Glossary/IIFE)
+in String format.
+The function will be invoked with three arguments: 
+the result of your function (an object with each parameter as a property), 
+a hint informing you of how the result will be shown to the user, 
+and a `Formatter` object. 
+The function can return a string, a formatted message object 
+(with the same structure as the JSON described here) 
+or an array of objects.
+
+The `Formatter` object provides the following methods for your convenience:
+- `measureToString(value, precision, unit)`: convert a value of `Measure` type to a string.
+- `dataToString(date)`: convert a value of `Date` type to a string containing only the date.
+- `timeToString(date)`: convert a value of `Date` type to a string containing only the time.
+- `dateAndTimeToString(date)`: convert a value of `Date` type to a string containing both the date and the time.
+- `locationToString(date)`: convert a value of `Location` type to a string
+- `anyToString(object)`: convert any other value to a string (unnecessary for `String` type)
+
+Here is an example used in [iCalendar](https://almond.stanford.edu/thingpedia/devices/by-id/org.thingpedia.icalendar).
+We want to offer different responses based on if the returned event contains `end_date` or not.
+The code looks like this:
+```javascript
+(function({start_date, end_date}, hint, formatter) {
+    if (end_date)
+        return `The event runs from ${formatter.dateAndTimeToString(start_date)} to ${formatter.dateAndTimeToString(end_date)}`;
+    else
+        return `The event starts at ${formatter.dateAndTimeToString(start_date)}`;
+})
+```
+And the annotation will look like this: 
+```tt
+#_[formatted=[{type="code",code="(function({start_date, end_date}, hint, formatter) {\nif (end_date)\nreturn `The event runs from ${formatter.dateAndTimeToString(start_date)} to ${formatter.dateAndTimeToString(end_date)}`;\nelse\nreturn `The event starts at ${formatter.dateAndTimeToString(start_date)}`;\n}"}]]
+```
```

Comment:
Note: for security reasons, formatting functions run sandboxed. You cannot access other APIs, and cannot use `require` to import other nodejs modules. If you need more complex computation, you should perform it in the query function and return the value as a declared parameter.